### ISSUE TITLE
Rifle and ambiguous or inaccurate mime types.

### ISSUE THREAD

**Issue by markus-bauer** [ranger/ranger#2865]
From what I understand, rfile tries to determine the mime type from the file extensions using python's mimetypes library.
If that method fails, rifle uses the external "file" program, which reports a more accurate type.
https://github.com/ranger/ranger/blob/1a007158cef1279df68d0e8567b126def6b68046/ranger/ext/rifle.py#L319

The first method is faster, but can obviously be wrong. And there are cases where extension-to-mime mapping is ambiguous.
For example, 3gp can be video or audio (https://developer.mozilla.org/en-US/docs/Web/Media/Formats/Containers#3gp).

---
Possible solutions

1. Allow the file to have multiple mime types. Matching of rifle conditions would then match any of the file's types. This would require that mimetypes actually reports multiple types. (Does it?)
Rifle rules could then be used like this:
```
# File has mime type ["audio/3gpp" , "video/3gpp"]
mime audio = audio_player
mime video = video_player
# NOTE: There should also be a way to match multiple types
```
This would essentially let you map two commands to a file that has both types.
The same is currently possible with extension:
```
ext 3gp = audio_player
ext 3gp = video_player
```
But it's not possible with multiple mime types, because the file only has one (AFAIU).

2. Have a new rifle condition to (try to) force an accurate mime type. If this condition is set, rifle would always try to use the more accurate method, and fall back to using the extension if that fails. The rule would be something like this:
```conf
# use the fastest mime method
ext jpg, mime image =  ...
# use the most accurate mime method
ext 3gp, accurate_mime video = ...
```
```conf
# mimetype by extensions says it has mime audio, so this wouldn't match
ext 3gp, mime video = ...
# "file" reports that this is actually a video, and this would match
ext 3gp, accurate_mime video = ...
```

3. Automatically use "file" to disambiguate. If file has multiple types, or if type is known to be ambiguous, directly use "file". Otherwise fall back to extension.
This could theoretically also be a condition, where it only disambiguates if the condition is set.
```conft
ext 3gp, mime audio = ...
ext 3gp, auto_disambiguate_mime video = ...
```

These are just some first thoughts to raise the issue.

**Comment by toonn** [comment]
Good idea but `mimetypes` can't do this. It canonicalizes the extension (`.tgz` -> `.tar.gz`) and then looks up the corresponding MIME in a dictionary. So there's only a single mapping possible. (Though just now when I opened `help(mimetypes)` in IDLE I could've sworn I saw `'.3gp': 'audio/3gpp'` in the `types_map` excerpt. Must've hallucinated that though because Python dictionaries can't have multiple values for a single key.)

TBH I don't think we should do anything more than making the use of `file(1)` configurable. Probably through a CLI option.

Is there even a convincing scenario where you might want to use a different program for something that might be either video or audio than when it is either of the two types individually? I can only come up with contrived examples.

I also don't think `xdg-open`, the most popular program that does a similar thing to Rifle, has such functionality? I'm not saying we should limit ourselves to what other programs do (what would even be the point of Rifle in that case?) but it does suggest maybe it's a solution looking for a problem.

**Comment by markus-bauer** [comment]
>Good idea but mimetypes can't do this.

That's an arbitrary limitation. Makes this library useless IMO, since these mappings are ambigous.
Okay, then having multiple mime types won't work.

> TBH I don't think we should do anything more than making the use of file(1) configurable. Probably through a CLI option.

I don't really understand what you mean by making the `file` call itself configurable.
I think you would need to have the settings in the rifle rules.

> Is there even a convincing scenario where you might want to use a different program for something that might be either video or audio than when it is either of the two types individually? I can only come up with contrived examples.

I'm also not sure what you mean here. 
There might be other ambiguous types. One would have to investigate this. Perhaps someone else can add examples.
Is the following example contrived?
```
# rifle rule
mime audio = audio_player
mime video = video_player
```
3gp video wouldn't play.
Extension mapping says audio, but "file" looks at the file's content (AFAIK) and is accurately saying video.
Rifle will always use extension and only uses `file` if the extension is unknown (which it's not for common types).

How can this be solved? One of my ideas was that you could force using `file`, with a flag or rule, only for the types you want. But that was just an idea.


###  Summary:
1) Rifle uses different (staged) methods to determine the mime type. In order of efficiency, but not in order of quality.
2) There's no way to tell rifle which methods should be used for a specific rule. I can't force a more accurate but slow method. Also, I'm saying `file`, but it could be any more accurate method used by rifle (now or in the future).

Again, I'm not sure this is the right solution. But the issue is real.

**Comment by markus-bauer** [comment]
Also, FWITW, xdg-open/xdg-mime actually report the correct type.
I don't know if you can configure the kind of matching it's doing.

I would have to look at the source code more deeply at some point. The relevant sections of xdg-mime, on my install, seems to be:
(Sorry for not finding a link to the source code. I don't have time right now)
```shell
info_kde()
{
    if [ -n "${KDE_SESSION_VERSION}" ]; then
      case "${KDE_SESSION_VERSION}" in
        4)
          DEBUG 1 "Running kmimetypefinder \"$1\""
          kmimetypefinder "$1" 2>/dev/null | head -n 1
        ;;
        5)
          DEBUG 1 "Running kmimetypefinder${KDE_SESSION_VERSION} \"$1\""
          kmimetypefinder${KDE_SESSION_VERSION} "$1" 2>/dev/null | head -n 1
        ;;
      esac
    else
        DEBUG 1 "Running kfile \"$1\""
        kfile "$1" 2> /dev/null | head -n 1 | cut -d "(" -f 2 | cut -d ")" -f 1
    fi

    if [ $? -eq 0 ]; then
        exit_success
    else
        exit_failure_operation_failed
    fi
}

info_gnome()
{
    if gio help info 2>/dev/null 1>&2; then
        DEBUG 1 "Running gio info \"$1\""
        gio info "$1" 2> /dev/null | grep standard::content-type | cut -d' ' -f4
    elif gvfs-info --help 2>/dev/null 1>&2; then
        DEBUG 1 "Running gvfs-info \"$1\""
        gvfs-info "$1" 2> /dev/null | grep standard::content-type | cut -d' ' -f4
    elif gnomevfs-info --help 2>/dev/null 1>&2; then
        DEBUG 1 "Running gnomevfs-info \"$1\""
        gnomevfs-info --slow-mime "$1" 2> /dev/null | grep "^MIME" | cut -d ":" -f 2 | sed s/"^ "//
    fi

    if [ $? -eq 0 ]; then
        exit_success
    else
        exit_failure_operation_failed
    fi
}

info_generic()
{
    if mimetype --version >/dev/null 2>&1; then
        DEBUG 1 "Running mimetype --brief --dereference \"$1\""
        mimetype --brief --dereference "$1"
    else
        DEBUG 1 "Running file --brief --dereference --mime-type \"$1\""
        /usr/bin/file --brief --dereference --mime-type "$1" 2> /dev/null
    fi

    if [ $? -eq 0 ]; then
        exit_success
    else
        exit_failure_operation_failed
    fi
}
```

**Comment by toonn** [comment]
Yeah, so `xdg-open` is basically just using `file(1)` or an equivalent. That's what I'm saying should be configurable, with an option, `rifle --use-magic`.

As for the example. I'd want to see a file that reasonably matches multiple MIME types and you can do something sensible with that information. In case of the 3gp file it's still either audio-only or video, no? Even if you can distinguish between audio-only, video-only and both, what would you do with that info, use a different video player when audio's included?

**Comment by toonn** [comment]
Let's keep this issue open as a "Implement an option for Rifle to always use `file(1)`" feature request.

**Comment by markus-bauer** [comment]
Final words on this:
I already stated why I think your solution is not the right way. I should be able to specify the method for specific rifle rules. Always on or always off could be a second option.
But you also made your point about why you think this isn't necessary. 
So I'm okay with not continuing right now.

**Comment by martin-braun** [comment]
On MacOS, there is also the `open` command that will use the program that is configured in their file browser `Finder`. Is there a way to tell `ranger` to just use `open`?
